docs/style/coding/CODING_STYLE_GUIDE.adoc

```diff
@@ -295,6 +296,368 @@ includes the header to also be using the namespace. This causes
+(POD)] data structures, this can be done by just allocating the
+structure at an appropriate scope. Alternatively, _raw_ storage can be
+allocated and then cast. However, great care must be taken with the
+latter approach to ensure that natural machine alignments and language
+strict-aliasing rules are observed. With the simple data structure
+declaration, the compiler does this on your behalf. With the raw
+approach, you must do this.
+
+Once the storage has been allocated, then use symmetric initializers and
+deinitializers such as those, for example, for `pthread_attr_t`. An
+example is shown in the listing below.
+
+[source,C,caption='',title='{listing-caption} *{counter:refnum}*. Using in place allocation and initialization in C or C{plusplus}.']
+----
+#include <pthread.h>
+#include <stdint.h>
+
+...
+
+#if USE_STRUCT_STORAGE
+// Allocate the structure directly.
+static pthread_attr_t sThreadAttributes;
+
+#elif USE_RAW_STORAGE
+// Allocate the structure using "raw" storage.
+#define chipDEFINE_ALIGNED_VAR(name, size, align_type) \
+    align_type name[(((size) + (sizeof(align_type) - 1)) / sizeof(align_type))]
+
+static chipDEFINE_ALIGNED_VAR(sThreadAttributes, sizeof (pthread_attr_t), uint64_t);
+
+#endif // USE_STRUCT_STORAGE
+
+int foobar(void)
+{
+    int              retval;
+    int              status;
+    pthread_t        thread;
+    pthread_attr_t * attrs = (pthread_attr_t *)&sThreadAttributes;
+
+    // Now "construct" or initialize the storage.
+    retval = pthread_attr_init(attrs);
+
+    if (retval == 0)
+    {
+        retval = pthread_create(&thread, attrs, NULL, NULL);
+
+        if (retval == 0)
+        {
+            status = pthread_join(thread, NULL);
+
+            if (status != 0)
+            {
+                retval = status;
+            }
+
+            status = pthread_attr_destroy(attrs);
+
+            if (status != 0)
+            {
+                retval = status;
+            }
+        }
+    }
+
+    return (retval);
+}
+----
+
+For non-scalar types and objects such as C{plusplus} classes, this gets slightly
+trickier since C{plusplus} constructors and destructors must be accounted for
+and invoked. Fortunately, C{plusplus} has placement new which handles this.
+The listing below modifies the listing above using C{plusplus} placement new
+to ensure the class is properly constructed before initialization and
+destructed after deinitialization.
+
+[source,C++,caption='',title='{listing-caption} *{counter:refnum}*. Using C{plusplus} placement new for in place allocation and initialization.']
+----
+#include <new>
+
+#include <pthread.h>
+#include <stdint.h>
+
+...
+
+class ThreadAttributes
+{
+public:
+    ThreadAttributes(void) {};
+    ~ThreadAttributes(void) {};
+
+    operator pthread_attr_t *(void) { return &mAttributes; }
+
+private:
+    pthread_attr_t mAttributes;
+};
+
+// Allocate the structure using "raw" storage.
+#define chipDEFINE_ALIGNED_VAR(name, size, align_type) \
```

Comment:
If we already require c++11, should this be

`alignas(align_type) unsigned char name[size];
`
or
```
#include <type_traits>
typename std::aligned_storage<size, alignof(align_type)>::type name;
```

?

We have standard library primitives for these patterns now.

Reply:
Makes sense to me; I'll update the listing.

Reply:
Done.